### PR TITLE
refactor(ivy): simplify logic of projectable nodes insertion

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
This PR is part of an ongoing refactoring effort to remove code duplication around projectable nodes handling - since we need to change data structures used by projectable nodes need to remove code duplication first.

This commit removes code duplication around projectable nodes insertion.
It also simplify the overall logic by using recursive function calls instead of hand-unrolled stack.

Note: we can always replace recursion with iteration but since this is an ongoing refactoring I would prefer to start with small / readable code and _then_ perf-optimised as needed. In fact I've prepared a commit with recursion -> iteration (see 38574ee2a) but I can see several problems with this approach:
- code is harder to follow;
- the actual code is longer;
- we need to have repeated Array.isArray checks (or duplicate code);
- it is way too easy to make a mistake in the custom recursion implementation
(in fact previous implementation had a bug where a stack was not cleaned up). 

tl;dr; I would prefer to err on the short / easy to follow code and optimise hotspots based on performance numbers.

### CODE REVIEW

**Review by kara — APPROVED** [review]
LGTM

**Comment by kara** [comment]
[presubmit](http://test/OCL:235973506:BASE:235975799:1551301260320:b4780e58)
